### PR TITLE
executor: support innodb_lock_wait_timeout  for pessimistic transaction

### PR DESCRIPTION
**Pull request by cfzjywxk**
### What problem does this PR solve? 

make `innodb_lock_wait_timeout` work for pessimistic transactions in tidb

in MySQL, this is a system variable to set lock wait time for transactions, 
The length of time in seconds an InnoDB transaction waits for a row lock before giving up. The default value is 50 seconds. A transaction that tries to access a row that is locked by another InnoDB transaction waits at most this many seconds for write access to the row before issuing the following error:
ERROR 1205 (HY000): Lock wait timeout exceeded; try restarting transaction
When a lock wait timeout occurs, the current statement is rolled back (not the entire transaction). 

[mysql-doc](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_lock_wait_timeout)

### What is changed and how it works?

- make system variable `innodb_lock_wait_timeout` effective，the default value is same with mysql 
- check wait time when met with `ErrLocked` error from kv, and report lock wait timout error if wait time greater than `lockWaitTime`, the current statement will be rollbacked like mysql behaviors
- `innodb_lock_wait_timeout` does affect only user sessions dmls

TODO:
1. refactor `handlePessimiticError`  to always update for `forUpdateTs`
2. make lock wait action in mvcc_leveldb same with action in tikv.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

Release note

 - Write release note for bug-fix or new feature.

### CODE REVIEW

File: session/pessimistic_test.go [hunk 1]
```diff
@@ -439,3 +567,65 @@ func (s *testPessimisticSuite) TestKillStopTTLManager(c *C) {
...
+
+	// tk2 lock c1 = 1
+	tk2.MustExec("begin pessimistic")
+	tk2.MustExec("select * from tk where c1 = 1 for update") // lock succ c1 = 1
+
+	// tk3 try lock c1 = 1 timeout 1sec
+	tk3.MustExec("begin pessimistic")
+	start := time.Now()
+	_, err := tk3.Exec("select * from tk where c1 = 1 for update")
+	c.Check(time.Since(start), GreaterEqual, time.Duration(1*time.Second))
```
Comment (coocood) [inline session/pessimistic_test.go]: We also need to check the duration is not too long.

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -447,6 +447,9 @@ type SessionVars struct {
 	isolationReadEngines map[kv.StoreType]struct{}
 
 	PlannerSelectBlockAsName []ast.HintTable
+
+	// Lock wait timeout for pessimistic transaction in milliseconds, `innodb_lock_wait_timeout` is in seconds
```
Comment (coocood) [inline sessionctx/variable/session.go]: put `is in seconds` at the last is too easy to be misunderstood as `LockWaitTimeout` is in seconds.

Comment (jackysp) [inline sessionctx/variable/session.go]:
```suggestion
	// LockWaitTimeout is used for pessimistic transaction in milliseconds, `innodb_lock_wait_timeout` is in seconds.
```

File: executor/executor.go [hunk 3]
```diff
@@ -815,7 +815,7 @@ func (e *SelectLockExec) Next(ctx context.Context, req *chunk.Chunk) error {
 		}
 		return nil
 	}
-	var lockWaitTime = kv.LockAlwaysWait
+	var lockWaitTime = e.ctx.GetSessionVars().LockWaitTimeout
```
Comment (jackysp) [inline executor/executor.go]:
```suggestion
	lockWaitTime := e.ctx.GetSessionVars().LockWaitTimeout
```

File: store/tikv/2pc.go [hunk 4]
```diff
@@ -726,15 +727,28 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
...
-				// the pessimistic lock found could be lock left behind(timeout but not recycled yet)
-				if !c.store.oracle.IsExpired(lock.TxnID, lock.TTL) {
-					return ErrLockAcquireFailAndNoWaitSet
+			if lock.LockType == pb.Op_PessimisticLock {
+				if action.lockWaitTime == kv.LockNoWait {
+					// the pessimistic lock found could be lock left behind(timeout but not recycled yet)
+					if !c.store.oracle.IsExpired(lock.TxnID, lock.TTL) {
+						return ErrLockAcquireFailAndNoWaitSet
+					}
+				} else if action.lockWaitTime == kv.LockAlwaysWait {
```
Comment (coocood) [inline store/tikv/2pc.go]: We don't need to handle `LockAlwaysWait`, `innodb_wait_timeout` is always positive.

Comment (jackysp) [inline store/tikv/2pc.go]: How about
```suggestion
				} else if action.lockWaitTime != kv.LockAlwaysWait {
```
then do something.

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -812,6 +816,9 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 	case MaxExecutionTime:
 		timeoutMS := tidbOptPositiveInt32(val, 0)
 		s.MaxExecutionTime = uint64(timeoutMS)
+	case InnodbLockWaitTimeout:
+		lockWaitSec := tidbOptInt64(val, DefInnodbLockWaitTimeout)
```
Comment (coocood) [inline sessionctx/variable/session.go]: We need to validate the range.

File: session/pessimistic_test.go [hunk 6]
```diff
@@ -567,3 +567,91 @@ func (s *testPessimisticSuite) TestKillStopTTLManager(c *C) {
...
+	tk2.MustQuery(`show variables like "innodb_lock_wait_timeout"`).Check(testkit.Rows("innodb_lock_wait_timeout 3"))
+	tk2.MustExec("set innodb_lock_wait_timeout = 2")
+	tk2.MustQuery(`show variables like "innodb_lock_wait_timeout"`).Check(testkit.Rows("innodb_lock_wait_timeout 2"))
+
+	tk3 := testkit.NewTestKitWithInit(c, s.store)
+	tk3.MustQuery(`show variables like "innodb_lock_wait_timeout"`).Check(testkit.Rows("innodb_lock_wait_timeout 3"))
+	tk3.MustExec("set innodb_lock_wait_timeout = 1")
+	tk3.MustQuery(`show variables like "innodb_lock_wait_timeout"`).Check(testkit.Rows("innodb_lock_wait_timeout 1"))
+
+	tk2.MustExec("set @@autocommit = 0")
```
Comment (jackysp) [inline session/pessimistic_test.go]: why set autocommit here?

File: store/tikv/2pc.go (hunk 4 above)
Comment (cfzjywxk) [reply store/tikv/2pc.go]: @coocood  some inner usages  which calls `doLockKeys`  (like `RecoverIndexExec`, `addIndexWorker`)   will set this param `LockAlwaysWait`

File: store/tikv/2pc.go [hunk 7]
```diff
@@ -676,6 +676,7 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
 		IsFirstLock:  c.isFirstLock,
 		WaitTimeout:  action.lockWaitTime,
 	}, pb.Context{Priority: c.priority, SyncLog: c.syncLog})
+	lockWaitStartTime := time.Now()
 	for {
```
Comment (coocood) [inline store/tikv/2pc.go]: We need to update the `action.lockWaitTime`. so if the wait timeout is less than 3s，we can return sooner.

File: sessionctx/variable/session.go (hunk 5 above)
Comment (cfzjywxk) [reply sessionctx/variable/session.go]: this is done in function `ValidateSetSystemVar`
```
case InnodbLockWaitTimeout:
	return checkUInt64SystemVar(name, value, 1, 1073741824, vars)
```

```
mysql> set global innodb_lock_wait_timeout = 0;
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+---------------------------------------------------------+
| Level   | Code | Message                                                 |
+---------+------+---------------------------------------------------------+
| Warning | 1292 | Truncated incorrect innodb_lock_wait_timeout value: '0' |
+---------+------+---------------------------------------------------------+
1 row in set (0.00 sec)
```

File: session/pessimistic_test.go (hunk 6 above)
Comment (cfzjywxk) [reply session/pessimistic_test.go]: set this before next `begin` statement  below

File: store/tikv/2pc.go [hunk 8]
```diff
@@ -726,15 +727,28 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
 			// if the lock left behind whose related txn is already committed or rollbacked,
 			// (eg secondary locks not committed or rollbacked yet)
 			// we cant return "nowait conflict" directly
-			if action.lockWaitTime == kv.LockNoWait && lock.LockType == pb.Op_PessimisticLock {
-				// the pessimistic lock found could be lock left behind(timeout but not recycled yet)
-				if !c.store.oracle.IsExpired(lock.TxnID, lock.TTL) {
-					return ErrLockAcquireFailAndNoWaitSet
+			if lock.LockType == pb.Op_PessimisticLock {
+				if action.lockWaitTime == kv.LockNoWait {
+					// the pessimistic lock found could be lock left behind(timeout but not recycled yet)
```
Comment (jackysp) [inline store/tikv/2pc.go]: It is better to improve the English of this comment.

File: store/tikv/2pc.go [hunk 9]
```diff
@@ -726,15 +727,28 @@ func (action actionPessimisticLock) handleSingleBatch(c *twoPhaseCommitter, bo *
...
+					if !c.store.oracle.IsExpired(lock.TxnID, lock.TTL) {
+						return ErrLockAcquireFailAndNoWaitSet
+					}
+				} else if action.lockWaitTime == kv.LockAlwaysWait {
+					// do nothing but keep wait
+				} else {
+					// user has set the `InnodbLockWaitTimeout`, check timeout
```
Comment (jackysp) [inline store/tikv/2pc.go]: It is better to improve the English of this comment.

File: store/tikv/2pc.go (hunk 4 above)
Comment (cfzjywxk) [reply store/tikv/2pc.go]: current we have 3 paths(negative, 0, positive value), I think it's more clear to make them like noodle style,  and we just need to add/remove one more `else if` path if we want to modify the protocol

File: store/tikv/2pc.go (hunk 7 above)
Comment (cfzjywxk) [reply store/tikv/2pc.go]: roger

**Comment by coocood** [comment]
LGTM

**Comment by cfzjywxk** [comment]
@tiancaiamao @lysu  PTAL

**Comment by coocood** [comment]
@youjiali1995 PTAL

**Pull request merged**